packages/react-import-remote/src/ImportRemote.tsx

```diff
@@ -21,91 +14,48 @@ export interface Props<Imported = any> {
   defer?: DeferTiming;
 }
 
-interface State {
-  loaded: boolean;
-  loading: boolean;
-}
-
-export default class ImportRemote extends React.PureComponent<Props, State> {
-  state: State = {loaded: false, loading: false};
-  private idleCallbackHandle?: RequestIdleCallbackHandle;
-
-  componentWillUnmount() {
-    if (this.idleCallbackHandle != null && 'cancelIdleCallback' in window) {
-      (window as WindowWithRequestIdleCallback).cancelIdleCallback(
-        this.idleCallbackHandle,
-      );
-    }
-  }
-
-  async componentDidMount() {
-    const {defer = DeferTiming.Mount} = this.props;
-
-    if (defer === DeferTiming.Idle) {
-      if ('requestIdleCallback' in window) {
-        this.idleCallbackHandle = (window as WindowWithRequestIdleCallback).requestIdleCallback(
-          this.loadRemote,
-        );
-      } else {
-        this.loadRemote();
+export default function ImportRemote({
+  source,
+  nonce,
+  preconnect,
+  onError,
+  getImport,
+  onImported,
+  defer,
+}: Props) {
+  const {result, intersectionRef} = useImportRemote(source, {
+    defer,
+    nonce,
+    getImport,
+  });
+
+  const intersectionObserver =
+    defer === DeferTiming.InViewport && intersectionRef ? (
+      <div ref={intersectionRef} />
+    ) : null;
+
+  React.useEffect(
+    () => {
+      switch (result.status) {
+        case Status.Failed:
+          onError(result.error);
+          return;
+        case Status.Complete:
+          onImported(result.imported);
       }
-    } else if (defer === DeferTiming.Mount) {
-      await this.loadRemote();
-    }
+    },
+    [result, onError, onImported],
```

Comment:
It is technically slightly more correct to split this across two effects, because this will cause `onImported` to run again even if only `onError` changed

Reply:
Although I wonder if it's better to just take a breaking change and change the callback to be a single callback that takes `Error | T`?